File: backend-cs/Services/MqttCommandHandler.cs
using System.Text;
using System.Text.Json;
using DriveChill.Hardware;
using Microsoft.Extensions.Logging;
using MQTTnet;
using MQTTnet.Client;

namespace DriveChill.Services;

/// <summary>
/// Background service that subscribes to MQTT command topics and dispatches
/// fan speed, profile activation, and fan release commands.
///
/// Allows external systems (Home Assistant, Node-RED) to control DriveChill via MQTT.
/// Creates separate subscriber clients (not shared with publish clients).
/// </summary>
public sealed class MqttCommandHandler : BackgroundService
{
    private readonly NotificationChannelService _channelSvc;
    private readonly FanService _fans;
    private readonly SettingsStore _store;
    private readonly IHardwareBackend _hw;
    private readonly ILogger<MqttCommandHandler> _log;

    private const int MaxCommandsPerSecond = 10;
    private const int ChannelCheckIntervalSeconds = 30;

    public MqttCommandHandler(
        NotificationChannelService channelSvc,
        FanService fans,
        SettingsStore store,
        IHardwareBackend hw,
        ILogger<MqttCommandHandler> log)
    {
        _channelSvc = channelSvc;
        _fans       = fans;
        _store      = store;
        _hw         = hw;
        _log        = log;
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        var activeTasks = new Dictionary<string, (CancellationTokenSource Cts, Task Task)>();

        try
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                try
                {
                    var channels = await _channelSvc.ListAsync(stoppingToken);
                    var currentIds = new HashSet<string>();

                    foreach (var ch in channels)
                    {
                        if (!ch.Enabled || ch.Type != "mqtt") continue;
                        if (!GetBool(ch.Config, "mqtt_subscribe", false)) continue;

                        currentIds.Add(ch.Id);

                        // Start or restart task if needed
                        if (!activeTasks.TryGetValue(ch.Id, out var entry) || entry.Task.IsCompleted)
                        {
                            entry.Cts?.Dispose();
                            var cts = CancellationTokenSource.CreateLinkedTokenSource(stoppingToken);
                            var task = SubscribeLoopAsync(ch.Config, ch.Name, cts.Token);
                            activeTasks[ch.Id] = (cts, task);
                        }
                    }

                    // Cancel tasks for removed/disabled channels
                    foreach (var id in activeTasks.Keys.ToList())
                    {
                        if (currentIds.Contains(id)) continue;
                        if (activeTasks.TryGetValue(id, out var old))
                        {
                            old.Cts.Cancel();
                            old.Cts.Dispose();
                            activeTasks.Remove(id);
                        }
                    }
                }
                catch (OperationCanceledException) when (stoppingToken.IsCancellationRequested)
                {
                    break;
                }
                catch (Exception ex)
                {
                    _log.LogDebug(ex, "MQTT command handler channel check error");
                }

                await Task.Delay(TimeSpan.FromSeconds(ChannelCheckIntervalSeconds), stoppingToken);
            }
        }
        finally
        {
            foreach (var (_, (cts, _)) in activeTasks)
            {
                cts.Cancel();
                cts.Dispose();
            }
        }
    }

    private async Task SubscribeLoopAsync(
        Dictionary<string, JsonElement> config,
        string channelName,
        CancellationToken ct)
    {
        var brokerUrl = GetStr(config, "broker_url") ?? "";
        if (string.IsNullOrEmpty(brokerUrl)) return;

        Uri uri;
        try { uri = new Uri(brokerUrl); }
        catch { return; }

        var hostname = uri.Host;
        var port = uri.Port > 0 ? uri.Port : (uri.Scheme == "mqtts" ? 8883 : 1883);
        var useTls = uri.Scheme is "mqtts" or "ssl";
        var username = GetStr(config, "username");
        var password = GetStr(config, "password");
        var topicPrefix = GetStr(config, "topic_prefix") ?? "drivechill";
        var clientId = $"drivechill-sub-{channelName[..Math.Min(8, channelName.Length)]}";

        var factory = new MqttFactory();
        using var client = factory.CreateMqttClient();

        var optionsBuilder = new MqttClientOptionsBuilder()
            .WithTcpServer(hostname, port)
            .WithClientId(clientId);

        if (useTls)
            optionsBuilder.WithTlsOptions(o => { });
        if (!string.IsNullOrEmpty(username))
            optionsBuilder.WithCredentials(username, password ?? "");

        // Rate tracking
        var rateTracker = new List<double>();
        var stopwatch = System.Diagnostics.Stopwatch.StartNew();

        // Set up message handler before connecting
        client.ApplicationMessageReceivedAsync += async e =>
        {
            // Rate limit
            var now = stopwatch.Elapsed.TotalSeconds;
            rateTracker.RemoveAll(t => now - t >= 1.0);
            if (rateTracker.Count >= MaxCommandsPerSecond)
            {
                _log.LogWarning("MQTT command rate limit exceeded for channel {Channel}, dropping", channelName);
                return;
            }
            rateTracker.Add(now);

            var topic = e.ApplicationMessage.Topic;
            var payload = e.ApplicationMessage.PayloadSegment.Count > 0
                ? Encoding.UTF8.GetString(e.ApplicationMessage.PayloadSegment)
                : null;

            try
            {
                DispatchCommand(topic, payload, topicPrefix);
            }
            catch (Exception ex)
            {
                _log.LogDebug(ex, "MQTT command dispatch error");
            }

            await Task.CompletedTask;
        };

        try
        {
            await client.ConnectAsync(optionsBuilder.Build(), ct);
            await client.SubscribeAsync(
                new MqttTopicFilterBuilder()
                    .WithTopic($"{topicPrefix}/commands/#")
                    .Build(),
                ct);

            _log.LogInformation("MQTT subscribed to {Prefix}/commands/# for channel {Channel}",
                topicPrefix, channelName);

            // Keep alive until cancelled
            while (!ct.IsCancellationRequested)
            {
                await Task.Delay(TimeSpan.FromSeconds(5), ct);
                if (!client.IsConnected) break;
            }
        }
        catch (OperationCanceledException) { /* expected on shutdown */ }
        catch (Exception ex)
        {
            _log.LogDebug(ex, "MQTT subscribe loop error for channel {Channel}", channelName);
        }
        finally
        {
            try
            {
                if (client.IsConnected)
                    await client.DisconnectAsync(new MqttClientDisconnectOptionsBuilder().Build());
            }
            catch { /* best-effort */ }
        }
    }

    /// <summary>
    /// Parse and dispatch a single MQTT command. Called from the message handler.
    /// </summary>
    internal void DispatchCommand(string topic, string? payload, string prefix)
    {
        var expectedStart = $"{prefix}/";
        if (!topic.StartsWith(expectedStart)) return;

        var suffix = topic[expectedStart.Length..];

        // commands/fans/release — no payload needed
        if (suffix == "commands/fans/release")
        {
            _fans.ReleaseFanControl();
            _log.LogInformation("MQTT: released fan control");
            return;
        }

        if (string.IsNullOrEmpty(payload))
        {
            _log.LogDebug("Empty MQTT payload for topic {Topic}", topic);
            return;
        }

        JsonElement data;
        try
        {
            data = JsonSerializer.Deserialize<JsonElement>(payload);
        }
        catch (JsonException)
        {
            _log.LogWarning("Malformed JSON on MQTT command topic {Topic}", topic);
            return;
        }

        // commands/fans/{fan_id}/speed
        if (suffix.StartsWith("commands/fans/") && suffix.EndsWith("/speed"))
        {
            var parts = suffix.Split('/');
            if (parts.Length == 4)
            {
                var fanId = parts[2];
                if (data.TryGetProperty("percent", out var percentEl)
                    && percentEl.ValueKind == JsonValueKind.Number)
                {
                    var percent = percentEl.GetDouble();
                    if (percent >= 0 && percent <= 100)
                    {
                        _hw.SetFanSpeed(fanId, percent);
                        _log.LogInformation("MQTT: set fan {FanId} to {Percent}%", fanId, percent);
                    }
                    else
                    {
                        _log.LogWarning("Invalid percent in MQTT fan speed command: {Percent}", percent);
                    }
                }
                else
                {
                    _log.LogWarning("Missing or invalid 'percent' in MQTT fan speed command");
                }
            }
        }
        // commands/profiles/activate
        else if (suffix == "commands/profiles/activate")
        {
            if (data.TryGetProperty("profile_id", out var pidEl)
                && pidEl.ValueKind == JsonValueKind.String)
            {
                var profileId = pidEl.GetString()!;
                var profiles = _store.LoadProfiles().ToList();
                var profile = profiles.FirstOrDefault(p => p.Id == profileId);
                if (profile != null)
                {
                    // Mark active and save
                    foreach (var p in profiles) p.IsActive = p.Id == profileId;
                    _store.SaveProfiles(profiles);
                    _fans.SetCurves(profile.Curves);
                    _log.LogInformation("MQTT: activated profile {ProfileId}", profileId);
                }
                else
                {
                    _log.LogWarning("MQTT: profile {ProfileId} not found", profileId);
                }
            }
            else
            {
                _log.LogWarning("Missing or invalid 'profile_id' in MQTT activate command");
            }
        }
        else
        {
            _log.LogDebug("Unknown MQTT command topic: {Topic}", topic);
        }
    }

    private static string? GetStr(Dictionary<string, JsonElement> config, string key)
        => config.TryGetValue(key, out var v) && v.ValueKind == JsonValueKind.String ? v.GetString() : null;

    private static bool GetBool(Dictionary<string, JsonElement> config, string key, bool defaultValue)
    {
        if (config.TryGetValue(key, out var v))
        {
            if (v.ValueKind is JsonValueKind.True or JsonValueKind.False) return v.GetBoolean();
            if (v.ValueKind == JsonValueKind.String && bool.TryParse(v.GetString(), out var parsed)) return parsed;
        }
        return defaultValue;
    }
}

File: backend/app/services/mqtt_command_handler.py
"""MQTT command handler — subscribes to command topics and dispatches actions.

Allows external systems (Home Assistant, Node-RED) to control DriveChill
via MQTT: set fan speeds, activate profiles, release fan override.

Runs as a background coroutine started from main.py lifespan.
"""
from __future__ import annotations

import asyncio
import json
import logging
import time
from typing import TYPE_CHECKING
from urllib.parse import urlparse

if TYPE_CHECKING:
    from app.services.notification_channel_service import NotificationChannelService
    from app.services.fan_service import FanService
    from app.db.repositories.profile_repo import ProfileRepo
    from app.hardware.base import HardwareBackend

logger = logging.getLogger(__name__)

MAX_COMMANDS_PER_SECOND = 10


async def _dispatch_command(
    topic: str,
    payload: bytes | bytearray | None,
    prefix: str,
    backend: "HardwareBackend",
    fan_service: "FanService",
    profile_repo: "ProfileRepo",
) -> None:
    """Parse and dispatch a single MQTT command."""
    # Strip the prefix to get the relative command path
    expected_start = f"{prefix}/"
    if not topic.startswith(expected_start):
        logger.debug("MQTT command topic does not match prefix: %s", topic)
        return

    suffix = topic[len(expected_start):]

    if not payload:
        # commands/fans/release needs no payload
        if suffix == "commands/fans/release":
            await fan_service.release_fan_control()
            logger.info("MQTT: released fan control")
            return
        logger.debug("Empty payload for %s", topic)
        return

    try:
        data = json.loads(payload)
    except (json.JSONDecodeError, ValueError):
        logger.warning("Malformed JSON on MQTT command topic %s", topic)
        return

    # commands/fans/{fan_id}/speed
    if suffix.startswith("commands/fans/") and suffix.endswith("/speed"):
        parts = suffix.split("/")
        if len(parts) == 4:  # commands / fans / {fan_id} / speed
            fan_id = parts[2]
            percent = data.get("percent")
            if (
                percent is not None
                and isinstance(percent, (int, float))
                and 0 <= percent <= 100
            ):
                success = await backend.set_fan_speed(fan_id, percent)
                logger.info("MQTT: set fan %s to %s%% (success=%s)", fan_id, percent, success)
            else:
                logger.warning("Invalid percent in MQTT fan speed command: %s", data)
        else:
            logger.debug("Unknown MQTT fan command structure: %s", suffix)

    # commands/profiles/activate
    elif suffix == "commands/profiles/activate":
        profile_id = data.get("profile_id")
        if profile_id and isinstance(profile_id, str):
            result = await profile_repo.activate(profile_id)
            if result:
                # Also apply the profile to the fan service so curves take effect
                profile = await profile_repo.get(profile_id)
                if profile:
                    await fan_service.apply_profile(profile)
                logger.info("MQTT: activated profile %s", profile_id)
            else:
                logger.warning("MQTT: profile %s not found", profile_id)
        else:
            logger.warning("Invalid profile_id in MQTT activate command: %s", data)

    # commands/fans/release (with payload — also accepted)
    elif suffix == "commands/fans/release":
        await fan_service.release_fan_control()
        logger.info("MQTT: released fan control")

    else:
        logger.debug("Unknown MQTT command topic: %s", topic)


async def _subscribe_loop(
    channel_config: dict,
    channel_name: str,
    backend: "HardwareBackend",
    fan_service: "FanService",
    profile_repo: "ProfileRepo",
) -> None:
    """Subscribe to one MQTT channel's command topics and dispatch actions."""
    try:
        import aiomqtt
    except ImportError:
        logger.warning(
            "aiomqtt is not installed — MQTT command subscription unavailable. "
            "Install it with: pip install aiomqtt"
        )
        return

    broker_url = channel_config.get("broker_url", "")
    if not broker_url:
        return

    parsed = urlparse(broker_url)
    hostname = parsed.hostname or "localhost"
    port = parsed.port or (8883 if parsed.scheme == "mqtts" else 1883)
    use_tls = parsed.scheme in ("mqtts", "ssl")
    username = channel_config.get("username") or None
    password = channel_config.get("password") or None
    topic_prefix = channel_config.get("topic_prefix", "drivechill")
    client_id = f"drivechill-sub-{channel_name[:8]}"

    tls_params = aiomqtt.TLSParameters() if use_tls else None

    rate_tracker: list[float] = []
    backoff = 5  # seconds, doubles on consecutive failures, caps at 60

    while True:
        try:
            async with aiomqtt.Client(
                hostname=hostname,
                port=port,
                username=username,
                password=password,
                identifier=client_id,
                tls_params=tls_params,
            ) as client:
                await client.subscribe(f"{topic_prefix}/commands/#")
                logger.info(
                    "MQTT subscribed to %s/commands/# for channel %s",
                    topic_prefix,
                    channel_name,
                )
                backoff = 5  # reset on successful connect

                async for message in client.messages:
                    # Rate limit: sliding window of 1 second
                    now = time.monotonic()
                    rate_tracker[:] = [t for t in rate_tracker if now - t < 1.0]
                    if len(rate_tracker) >= MAX_COMMANDS_PER_SECOND:
                        logger.warning(
                            "MQTT command rate limit exceeded for channel %s, dropping message",
                            channel_name,
                        )
                        continue
                    rate_tracker.append(now)

                    topic = str(message.topic)
                    try:
                        await _dispatch_command(
                            topic, message.payload, topic_prefix,
                            backend, fan_service, profile_repo,
                        )
                    except Exception:
                        logger.debug("MQTT command dispatch error", exc_info=True)
        except asyncio.CancelledError:
            raise  # let the supervisor cancel us cleanly
        except Exception:
            logger.warning(
                "MQTT subscribe loop error for channel %s, retrying in %ds",
                channel_name, backoff, exc_info=True,
            )
            await asyncio.sleep(backoff)
            backoff = min(backoff * 2, 60)


async def create_mqtt_command_handler(
    channel_svc: "NotificationChannelService",
    backend: "HardwareBackend",
    fan_service: "FanService",
    profile_repo: "ProfileRepo",
) -> None:
    """Main loop: watch for mqtt_subscribe-enabled channels, subscribe to each.

    Periodically re-checks channel list. Creates a separate aiomqtt client
    per channel (subscribe clients should be separate from publish clients).
    Cancelled tasks for removed/disabled channels are cleaned up automatically.
    """
    active_tasks: dict[str, asyncio.Task] = {}  # channel_id -> task

    try:
        while True:
            try:
                channels = await channel_svc.list_channels()
            except Exception:
                logger.debug("MQTT command handler: failed to list channels", exc_info=True)
                await asyncio.sleep(30)
                continue

            current_ids: set[str] = set()

            for ch in channels:
                if not ch.enabled or ch.type != "mqtt":
                    continue
                if not ch.config.get("mqtt_subscribe", False):
                    continue

                current_ids.add(ch.id)

                # Start or restart task if needed
                if ch.id not in active_tasks or active_tasks[ch.id].done():
                    task = asyncio.create_task(
                        _subscribe_loop(
                            ch.config, ch.name,
                            backend, fan_service, profile_repo,
                        )
                    )
                    active_tasks[ch.id] = task

            # Cancel tasks for removed/disabled channels
            for cid in list(active_tasks):
                if cid not in current_ids:
                    active_tasks[cid].cancel()
                    del active_tasks[cid]

            await asyncio.sleep(30)  # Re-check channels every 30s
    except asyncio.CancelledError:
        for task in active_tasks.values():
            task.cancel()
        await asyncio.gather(*active_tasks.values(), return_exceptions=True)
